### PR TITLE
Migrate the S3 SDK from v1 to v2

### PR DESCRIPTION
**Pull request by frouioui**
## Description

This PR updates the version of the AWS SDK from v1 to v2. The version 1 has been deprecated for some time now.

I have ran the following manual test:
1. Create a Vitess cluster (1 unsharded keyspace) from the `main` branch
1. Insert some data
1. Backup the cluster into AWS S3
1. Teardown the cluster
1. Create a Vitess cluster (same config) from the `migrate-to-s3-v2` branch (with `--restore_from_backup`)
1. Restore from the previous backup
1. Observe that the inserted data shows up after restoring
1. Take another backup on AWS S3
1. List all available backup on AWS S3
2. Restore the cluster from the latest backup (using `RestoreFromBackup`)
1. Remove the first backup we did (the one from `main`) from AWS S3
2. Teardown the cluster
3. Create another Vitess cluster (same config) from the `main` branch (with `--restore_from_backup`)
4. Observe that the inserted data shows up after restoring

This test ensures that we can upgrade: old backups will still work with the new code, and that we can downgrade: new backups will work with the old code. While making sure that no config/flag change is needed between the two code version. It also checks that the `GetBackups` and `RestoreFromBackup` vtctldclient commands work.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/mysqlctl/s3backupstorage/s3.go [hunk 1]
```diff
@@ -169,25 +177,32 @@ func (bh *S3BackupHandle) AddFile(ctx context.Context, filename string, filesize
...
 		defer bh.waitGroup.Done()
-		uploader := s3manager.NewUploaderWithClient(bh.client, func(u *s3manager.Uploader) {
+		uploader := manager.NewUploader(bh.client, func(u *manager.Uploader) {
 			u.PartSize = partSizeBytes
 		})
 		object := objName(bh.dir, bh.name, filename)
 		sendStats := bh.bs.params.Stats.Scope(stats.Operation("AWS:Request:Send"))
 		// Using UploadWithContext breaks uploading to Minio and Ceph https://github.com/vitessio/vitess/issues/14188
-		_, err := uploader.Upload(&s3manager.UploadInput{
+		_, err := uploader.Upload(context.Background(), &s3.PutObjectInput{
```
Comment (frouioui) [inline go/vt/mysqlctl/s3backupstorage/s3.go:186]: Using `context.Background()` is a side-effect of the issue described in https://github.com/vitessio/vitess/issues/14188. Upgrading to the SDK V2 does not fix it, the issue is rooted in how we handle the context and the goroutines. I started investigating the issue a bit deeper and will continue in a subsequent PR.

File: go/vt/mysqlctl/s3backupstorage/retryer_test.go [hunk 2]
```diff
@@ -1,79 +1,69 @@
...
 		fallbackPolicy bool
 		expected       bool
 	}{
-
 		{
-			name: "non retryable request",
-			r: &request.Request{
-				Retryable: aws.Bool(false),
-			},
+			name:           "no error",
```
Comment (deepthi) [inline go/vt/mysqlctl/s3backupstorage/retryer_test.go:48]: Nice refactor.

File: go/vt/mysqlctl/xtrabackupengine.go [hunk 3]
```diff
@@ -277,7 +277,7 @@ func (be *XtrabackupEngine) executeFullBackup(ctx context.Context, params Backup
 	if err != nil {
 		return BackupUnusable, vterrors.Wrapf(err, "cannot JSON encode %v", backupManifestFileName)
 	}
-	if _, err := mwc.Write([]byte(data)); err != nil {
+	if _, err := mwc.Write(data); err != nil {
```
Comment (deepthi) [inline go/vt/mysqlctl/xtrabackupengine.go:280]: This code is going to be called regardless of s3 or other storage backend. Why this change?

Comment (frouioui) [reply go/vt/mysqlctl/xtrabackupengine.go:280]: `data` is a already a slice of bytes

**Review by deepthi — APPROVED** [review]

**Review by mattlord — APPROVED** [review]

**Comment by marf** [comment]
Hi, I have recently upgraded to Vitess v21 from v20. Now the backup fails with the following error:

```
+ /vt/bin/vtctldclient --action_timeout=2h0m0s --server esound-vtctld-d7707a8c.default.svc.cluster.local:15999 BackupShard esound/-
E1130 13:16:50.577523       7 main.go:56] rpc error: code = Unknown desc = TabletManager.Backup on uk-0828527281: StartBackup failed: operation error S3: HeadBucket, resolve auth scheme: resolve endpoint: endpoint rule error, Custom endpoint `` was not a valid URI
```

I haven't changed how the backup configuration works and the s3 credentials and configuration are correct, I copy it here:

```
spec:
  backup:
    locations:
      - s3:
          authSecret:
            key: credentials
            name: vitess-s3-credentials
          bucket: esound-vitess-backup
          region: eu-west-2
```

Any idea what this could be?  Seems more an issue with this PR.

Thank you!

**Comment by marf** [comment]
I think I have found the issue.

Seems a bug with the fact that it does not check if the endpoint config is not specified and if not specified it uses an empty value instead of completely ignoring it and constructing the URL from the other parameters.

I managed to get it working by specifying this in the spec:

```
spec:
  backup:
    locations:
      - s3:
          authSecret:
            key: credentials
            name: vitess-s3-credentials
          bucket: esound-vitess-backup
          region: eu-west-2
          endpoint: https://s3.eu-west-2.amazonaws.com/ # fix for Vitess v21.0.0
```

**Comment by frouioui** [comment]
Hi @marf, thanks for raising this. I opened a direct Pull Request to `release-21.0` to fix the issue, we should be releasing a new v21 patch tonight. https://github.com/vitessio/vitess/pull/17307

**Comment by marf** [comment]
Thank you @frouioui.

As soon as released I will test it.

I want to raise also the fact that even in the case I specified a custom endpoint to make the backup working after a certain point during the finalization of the backup for a large table that we have called usertrack we get this error:

```
esound/- (uk-0828527281): time:{seconds:1733007057 nanoseconds:802234680} file:"builtinbackupengine.go" line:803 value:"backing up  \"vt_esound/userstracks.ibd\": 98.91% (417421408.00/422031360.00kb)"
esound/- (uk-0828527281): time:{seconds:1733007062 nanoseconds:802172059} file:"builtinbackupengine.go" line:803 value:"backing up  \"vt_esound/userstracks.ibd\": 99.38% (419396064.00/422031360.00kb)"
esound/- (uk-0828527281): time:{seconds:1733007067 nanoseconds:802024811} file:"builtinbackupengine.go" line:803 value:"backing up  \"vt_esound/userstracks.ibd\": 99.96% (421879616.00/422031360.00kb)"
esound/- (uk-0828527281): time:{seconds:1733007068 nanoseconds:95660900} file:"builtinbackupengine.go" line:902 value:"closing compressor"
esound/- (uk-0828527281): time:{seconds:1733007068 nanoseconds:97128328} file:"builtinbackupengine.go" line:795 value:"Completed backing up  \"vt_esound/userstracks.ibd\""
esound/- (uk-0828527281): time:{seconds:1733007081 nanoseconds:110832416} file:"builtinbackupengine.go" line:509 value:"resetting mysqld super_read_only to true"
esound/- (uk-0828527281): time:{seconds:1733007081 nanoseconds:111730683} file:"builtinbackupengine.go" line:526 value:"restarting mysql replication"
E1130 22:51:23.243289       7 main.go:56] rpc error: code = Unknown desc = TabletManager.Backup on uk-0828527281: upload multipart failed, upload id: PbdDGKgIlKtHUj0OYGkB.gI.g.TINRpUZtUZwHqOHkhthC9Uu04ScDsqiPDN5XMnvxkQW8T4d0Z01gpM2P26ebuHN73TVLYfa4pUI7I7Pg7bILDB2g13j0371Ga9KFKi, cause: operation error S3: UploadPart, https response error StatusCode: 0, RequestID: , HostID: , canceled, context canceled
```

I then downgraded to vitess v20.0.3 and this issue seems not to happen with this version. Not sure if it is something related with the upgrade of the S3 SDK.

**Comment by frouioui** [comment]
This is fixed by #17285 and is already included in the upcoming 21.0.1 release.

**Comment by frouioui** [comment]
> As soon as released I will test it.

Thanks for that! If you see any issue, let me know :)
